File: hw/quiroga/U3/HW_37_Strategy/StrategyExample/src/main/java/ec/edu/espe/strategyexample/model/QuickSort.java

package ec.edu.espe.strategyexample.model;

/**
 *
 * @author Esteban Quiroga 
 */
public class QuickSort implements SortingStrategy {

    @Override
    public int[] sort(int[] data) {
        int[] sortedData = data.clone(); // Avoid modifying the original array
        quickSortHelper(sortedData, 0, sortedData.length - 1);
        return sortedData;
    }
    
    private static void quickSortHelper(int[] data, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(data, low, high);
            quickSortHelper(data, low, pivotIndex - 1);
            quickSortHelper(data, pivotIndex + 1, high);
        }
    }

    private static int partition(int[] data, int low, int high) {
        int pivot = data[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (data[j] < pivot) {
                i++;
                // Swap
                int temp = data[i];
                data[i] = data[j];
                data[j] = temp;
            }
        }
        // Swap pivot to correct position
        int temp = data[i + 1];
        data[i + 1] = data[high];
        data[high] = temp;
        return i + 1;
    }
}

File: hw/pazmino/u3/HW37StrategyExercise/hw38StrategySortAppPython/model/quick_sort.py
from model.sorting_strategy import SortingStrategy

class QuickSort(SortingStrategy):
    def sort(self, data):
        print("Using QuickSort...")
        self.quick_sort(data, 0, len(data) - 1)
        return data

    def quick_sort(self, arr, low, high):
        if low < high:
            pivot_index = self.partition(arr, low, high)
            self.quick_sort(arr, low, pivot_index - 1)
            self.quick_sort(arr, pivot_index + 1, high)

    def partition(self, arr, low, high):
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] <= pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]  # Swap
        arr[i + 1], arr[high] = arr[high], arr[i + 1]  # Swap
        return i + 1
